modules/core/src/umatrix.cpp

```diff
@@ -637,18 +639,28 @@ Mat UMat::getMat(int accessFlags) const
 {
     if(!u)
         return Mat();
+    CV_Assert(!u->tempUMat() && "Don't get Mat from temp UMat! Use copyTo().");
     // TODO Support ACCESS_READ (ACCESS_WRITE) without unnecessary data transfers
     accessFlags |= ACCESS_RW;
-    u->currAllocator->map(u, accessFlags);
-    CV_Assert(u->data != 0);
-    Mat hdr(dims, size.p, type(), u->data + offset, step.p);
-    hdr.flags = flags;
-    hdr.u = u;
-    hdr.datastart = u->data;
-    hdr.data = u->data + offset;
-    hdr.datalimit = hdr.dataend = u->data + u->size;
-    CV_XADD(&hdr.u->refcount, 1);
-    return hdr;
+    UMatDataAutoLock autolock(u);
+    if(CV_XADD(&u->refcount, 1) == 0)
+        u->currAllocator->map(u, accessFlags);
+    if (u->data != 0)
+    {
+        Mat hdr(dims, size.p, type(), u->data + offset, step.p);
+        hdr.flags = flags;
+        hdr.u = u;
+        hdr.datastart = u->data;
+        hdr.data = u->data + offset;
+        hdr.datalimit = hdr.dataend = u->data + u->size;
+        return hdr;
+    }
+    else
+    {
+        CV_XADD(&u->refcount, -1);
+        CV_Assert(u->data != 0);
```

Comment:
Can we add some useful error description here (like above)? Or is it an "internal error" and should never happen?
```
CV_Assert(u->data != 0 && "Some useful message about that is going wrong and how to fix this");
```